openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart

```diff
@@ -148,6 +168,234 @@ class Openapi {
   });
 }
 
+/// Provides the input spec file to be used.
+///
+/// Provides the location of the input spec file to be used by the generator.
+/// Includes the option to use the default json or yaml paths.
+class InputSpec {
+  final String path;
+  final bool defaultYaml;
+
+  const InputSpec({String? path, this.defaultYaml = true})
+      : path = path ?? 'openapi.${defaultYaml ? 'yaml' : 'json'}';
+
+  const InputSpec.empty() : this();
+
+  const InputSpec.emptyJson() : this(defaultYaml: false);
+
+  Map<String, dynamic> toJsonMap() =>
+      {'path': path, 'defaultYaml': defaultYaml};
+
+  InputSpec.fromMap(Map<String, dynamic> map)
+      : this(
+          path: map['path'],
+          defaultYaml: map['defaultYaml'] == 'true' ? true : false,
+        );
+}
+
+/// Provides the location for the remote specification.
+///
+/// Provides basic support for fetching remote specification files hidden behind
+/// an authenticated endpoint.
+///
+/// By default when no [url] is provided a default [Uri.http] to
+/// localhost:8080/PWD is used.
+///
+/// This contains authentication information for fetching the OAS spec ONLY. This
+/// does not apply security to the entry points defined in the OAS spec.
+class RemoteSpec extends InputSpec {
+  final String? authHeaderContent;
+
+  const RemoteSpec({
+    required super.path,
+    this.authHeaderContent,
+  });
+
+  const RemoteSpec.empty() : this(path: 'http://localhost:8080/');
+
+  Uri get url => Uri.parse(path);
+
+  Map<String, String> toHeaderMap() {
+    return {
+      if (authHeaderContent != null)
+        'Authorization': 'Bearer ${authHeaderContent}',
+    };
+  }
+
+  Map<String, dynamic> toJsonMap() {
+    return {
+      if (authHeaderContent != null) 'authHeaderContent': authHeaderContent!,
+      ...super.toJsonMap(),
+    };
+  }
+
+  RemoteSpec.fromMap(Map<String, dynamic> map)
+      : authHeaderContent = map['authHeaderContent'],
+        super.fromMap(map);
+}
+
+/// Indicates whether or not the spec file live within AWS.
+///
+/// Since AWS handles the authentication header differently, we need to inform
+/// the builder to include the alternate auth header.
+///
+/// This currently only support AWS S3.
+///
+/// This contains authentication information for fetching the OAS spec ONLY. This
+/// does not apply security to the entry points defined in the OAS spec.
+class AwsRemoteSpec extends RemoteSpec {
+  /// The accessKeyId use to interact with AWS.
```

Comment:
I think Remotespec should provide a ```header``` property instead of dealing with the implementation details of composing an auth header for AWS and other remote servers. This way, the consumer deals with the edge cases themselves (environments, etc). Then there will be no need of creating ```AwsRemoteSpec```,``` GCPRemoteSpec``` etc

Reply:
Sorry for the slow reply. 🤔 I was tossing up between using a map vs the token string. One concern I have with placing the header generation on the user is the requirements that AWS has specifically. The request needs to be signed within a 15 minute window of being used and this will cause them to need to manually regenerate that header, thus requiring updates to the configuration. 

I may have a different use case where my flow would be to rarely update the Annotation but define the remote spec provider. While this does impose a bit on the library for the implementation it allows for the builder to properly handle the communication w/o requiring too much tweaking and manual changes from the developer. I personal would greatly dislike having to regenerate the signed header each and every time I want to regenerate the spec.

As for a  `GRPCRemoteSpec` I don't know enough about what is needed to even support the authentication story for that. 

Proposal:
- I change the RemoteSpec to take in headers
- We leave the AWS implementation so that the end user doesn't need to rebuild and sign the request
  - I can take ownership of the AWS implementation and thus when issues arise about it I will address the issues.
  - We only support the `GET` request to `S3` and we can take contributions for `GRPCRemoteSpec` or Azure etc with the explicit understanding they are responsible for both maintenance and triage. If it gets ignored by the contributor we will deprecate and drop support. 
    -  This may seem harsh but the support shouldn't need to fall on those that do not have an understanding of the subject matter.
    - This would require some level of staleness & community guidelines.
    - IIRC `GRPCRemoteSpec` would require that we have it installed into the repo and would require a larger lift due to the nature of GRPC, and this is likely unfeasible.

Reply:
I don't mean the user will generate the header manually. 

What I mean is that they will provide the implementation that generates the header.

Let elaborate.

If `RemoteSpec` has a field `hearderDelegate` of type `OpenapiHeaderDelegate`. Now, this  `OpenapiHeaderDepegate` can be extended by the user for whatever remote source they use for their spec.

This of course does not prevent the library from providing a default implementation like `AWSHeaderDelegate` or `MapHeaderDelegate`. I think this is more maintainable. The last thing we want is dropping support for useful features without providing alternatives for existing users.

This is just an example. And the code makes sense in my head. lol. 

What do you think?

Reply:
🤔 That makes sense to me. I misunderstood before. 

Would you mind expanding a bit more on what the delegate might look like?

Reply:
Sorry for late reply, been out of power. Something like 

```dart
///The header provider base class
class RemoteSpecHeaderDelegate{
  const RemoteSpecHeaderDelegate();
  Map<String,String>? getHeader(){
    return null;
  }
}

/// Default header provider provided by library
class AWSRemoteSpecHeaderDelegate extends RemoteSpecHeaderDelegate{
  
  Map<String,String>? getHeader(){
    return null;
  }

 const  AWSRemoteSpecHeaderDelegate();
}

/// header provider implemented by user
class UserImplementedHeaderDelegate extends RemoteSpecHeaderDelegate{

  const UserImplementedHeaderDelegate();

  @override
  Map<String, String>? getHeader() {
    return {'key':'value'};
  }
}

```

the below field is added to the openapi annotation or to the RemoteSpec class (preferrably). For example below, I added it to the openapi annotation

```dart
final RemoteSpecHeaderDelegate? headerDelegate;
```

and used like 

```dart
@Openapi(
  additionalProperties:
      DioProperties(pubName: 'petstore_api', pubAuthor: 'Johnny dep..'),
  inputSpecFile: 'openapi-spec.yaml',
  typeMappings: {'Pet': 'ExamplePet'},
  generatorName: Generator.dio,
  runSourceGenOnOutput: true,
  headerDelegate: UserImplementedHeaderDelegate(), // Header provider passed to the openapi annotation
  useNextGen: true,
  outputDirectory: 'api/petstore_api',
)
```

Reply:
on a second thought, `headerProvider` is probably a better name or just `header`

Reply:
I can try to explain better if you do not understand